heron/schedulers/src/java/com/twitter/heron/scheduler/mesos/framework/MesosFramework.java

```diff
@@ -0,0 +1,541 @@
+
+    return true;
+  }
+
+  // Restart a topology
+  public boolean restartJob(int containerIndex) {
+    if (isTerminated) {
+      LOG.severe("Job has been killed");
+      return false;
+    }
+
+    // List of tasks to restart
+    List<String> tasksToRestart = new ArrayList<>();
+    if (containerIndex == -1) {
+      // Restart all tasks
+      tasksToRestart.addAll(tasksId.values());
+    } else {
+      tasksToRestart.add(tasksId.get(containerIndex));
+    }
+
+    // Kill the task -- the framework will automatically restart it
+    for (String taskId : tasksToRestart) {
+      driver.killTask(Protos.TaskID.newBuilder().setValue(taskId).build());
+    }
+
+    return true;
+  }
+
+  /**
+   * Check whether the MesosFramework has been terminated
+   *
+   * @return true if MesosFramework has been terminated
+   */
+  public boolean isTerminated() {
+    return isTerminated;
+  }
+
+  /**
+   * Causes the current thread to wait for MesosFramework got registered,
+   * unless the thread is interrupted, or the specified waiting time elapses.
+   *
+   * @param timeout the maximum time to wait
+   * @param unit the time unit of the timeout argument
+   * @return true MesosFramework got registered,
+   * and false if the waiting time elapsed before the count reached zero
+   */
+  public boolean waitForRegistered(long timeout, TimeUnit unit) {
+    try {
+      if (this.registeredLatch.await(timeout, unit)) {
+        return true;
+      }
+    } catch (InterruptedException e) {
+      LOG.severe("Failed to wait for mesos framework got registered");
+      return false;
+    }
+
+    return false;
+  }
+
+  /**
+   * Get the FrameworkID registered with Mesos Master
+   *
+   * @return the FrameworkID
+   */
+  public Protos.FrameworkID getFrameworkId() {
+    return frameworkId;
+  }
+
+
+  @Override
+  public void registered(SchedulerDriver schedulerDriver,
+                         Protos.FrameworkID frameworkID,
+                         Protos.MasterInfo masterInfo) {
+    LOG.info("Registered! ID = " + frameworkID.getValue());
+    this.driver = schedulerDriver;
+    this.frameworkId = frameworkID;
+
+    // Release the latch
+    this.registeredLatch.countDown();
+  }
+
+  @Override
+  public void reregistered(SchedulerDriver schedulerDriver, Protos.MasterInfo masterInfo) {
+    LOG.info("Re-registered!");
+  }
+
+  @Override
+  public void resourceOffers(SchedulerDriver schedulerDriver, List<Protos.Offer> offers) {
+    LOG.fine("Received Resource Offers: " + offers.toString());
+
+    Map<Protos.Offer, TaskResources> offerResources = new HashMap<>();
+    // Convert the offers into TasksResources, an utils class.
+    for (Protos.Offer offer : offers) {
+      offerResources.put(offer, TaskResources.apply(offer, Context.role(heronConfig)));
+    }
+
+    // Generate launchable tasks
+    List<LaunchableTask> tasksToLaunch = generateLaunchableTasks(offerResources);
+
+    // Launch tasks
+    launchTasks(tasksToLaunch);
```

Comment:
If a task to be launched fails, what should we do with the resource offers? Should we decline the offer explicitly or we retry until the task is launched successfully?